### PR TITLE
Fix KMAC bounds checks

### PR DESCRIPTION
**Pull request by slontis**
Setting an output length higher than 8191 was causing a buffer overflow.
This was reported by Acumen (FIPS lab).
The max output size has increased to ~2M and it now checks this during set_parameters.
    
The encoder related functions now pass in the maximum size of the output buffer so they
can correctly check their size. kmac_bytepad_encode_key() calls bytepad twice in
order to calculate and check the length before encoding.
    
Note that right_encode() is currently only used in one place but this
may change if other algorithms are supported (such as TupleHash).

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: providers/common/provider_err.c [hunk 1]
```diff
@@ -112,7 +114,7 @@ static const ERR_STRING_DATA PROV_str_reasons[] = {
     "key size too small"},
     {ERR_PACK(ERR_LIB_PROV, 0, PROV_R_LENGTH_TOO_LARGE), "length too large"},
     {ERR_PACK(ERR_LIB_PROV, 0, PROV_R_MISMATCHING_DOMAIN_PARAMETERS),
-    "mismatching shared parameters"},
+    "mismatching domain parameters"},
```
Comment (slontis) [inline providers/common/provider_err.c:117]: oops : this is a left over problem from another PR.. I must of just done a search and replace. This change should go in..

File: providers/implementations/macs/kmac_prov.c [hunk 2]
```diff
@@ -246,15 +252,15 @@ static int kmac_setkey(struct kmac_data_st *kctx, const unsigned char *key,
...
+    if (keylen < KMAC_MIN_KEY || keylen > KMAC_MAX_KEY) {
         ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
         return 0;
     }
     if (w < 0) {
         ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST_LENGTH);
         return 0;
     }
-    if (!kmac_bytepad_encode_key(kctx->key, &kctx->key_len,
+    if (!kmac_bytepad_encode_key(kctx->key, sizeof(kctx->key), &kctx->key_len,
```
Comment (richsalz) [inline providers/implementations/macs/kmac_prov.c:263]: you're assuming the `sizeof` here, and below, have no padding.

File: providers/implementations/macs/kmac_prov.c [hunk 3]
```diff
@@ -483,17 +494,19 @@ static int right_encode(unsigned char *out, size_t *out_len, size_t bits)
  * e.g- in="KMAC" gives out[6] = { 0x01, 0x20, 0x4B, 0x4D, 0x41, 0x43 }
  *                                 len   bits    K     M     A     C
  */
-static int encode_string(unsigned char *out, size_t *out_len,
+static int encode_string(unsigned char *out, size_t out_max_len, size_t *out_len,
                          const unsigned char *in, size_t in_len)
 {
     if (in == NULL) {
         *out_len = 0;
```
Comment (richsalz) [inline providers/implementations/macs/kmac_prov.c:501]: could you just add a return here and drop the else and its indent?

Comment (slontis) [reply providers/implementations/macs/kmac_prov.c:501]: I dont think this is any cleaner for this case.

File: providers/implementations/macs/kmac_prov.c (hunk 2 above)
Comment (slontis) [reply providers/implementations/macs/kmac_prov.c:263]: Not quite sure what you are saying.
I am assuming sizeof(unsigned char) is 1., and that sizeof() returns the number of elements in the array that can be used. This is used quit a lot in the code base.

Comment (slontis) [reply providers/implementations/macs/kmac_prov.c:263]: NOTE key is this: unsigned char key[KMAC_MAX_KEY_ENCODED];

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Please note this will require make update (with enable-fips) after rebasing before merging.

**Comment by slontis** [comment]
Rebased and updated checksums. (Glad we have another file like libcrypto.num now :))

**Comment by slontis** [comment]
Thanks for the reminder about make update..

Merged to master

**Comment by kaduk** [comment]
>  and it now checks this during set_parameters.

Do we think we might have any other `set_parameters` that fail to check their respective lengths?

**Comment by paulidale** [comment]
Probably :(

**Pull request closed without merging**